add analyzer and code fixer for CA2225: Operator overloads have named alternates

brettfo opened the pull request:
This analyzer/fixer ensures that operator overloads have friendly alternate names.  E.g., `operator +` requires the presence of a public method named `Add`.

In this version I attempted to stay as close to the original FxCop implementation as possible, where any public method with the appropriate name would satisfy the requirement.  I'm open to ensuring the friendly-named method meet other requirements (e.g., static/instance, parameters, etc.) but I wanted some input on that, first, particularly around style.  If overloading `operator +` requires a method named `Add`, should that be a static method taking two parameters with a signature matching the operator overload, or an instance method taking just one?  What if there's a method named `Add` that takes four parameters?  What then?

For the fixer, operators that dictate the presence of a `CompareTo` or `ToInt32` (or `To`+someType) method, or operators that require an `IsTrue` property, those are created as instance methods/properties, where all other methods are created as static.

Tagging @srivatsn @mavasani @heejaechang @lgolding for review.

Fixes #541.

srivatsn commented on src/Microsoft.ApiDesignGuidelines.Analyzers/Core/OperatorOverloadsHaveNamedAlternates.Fixer.cs:
```diff
@@ -27,10 +32,126 @@ public sealed override FixAllProvider GetFixAllProvider()
+
+        private static void AddFriendlyAlternate(CodeFixContext context, SyntaxNode root, SyntaxNode node, SemanticModel semanticModel, Diagnostic diagnostic)
+        {
+            var operatorOverload = (SyntaxNode)node.FirstAncestorOrSelf<TOperatorDeclarationSyntax>() ?? node.FirstAncestorOrSelf<TConversionOperatorSyntax>();
```
you can simply call `syntaxGenerator.GetDeclaration(node, DeclarationKind.Operator)` to get the enclosing operator declaration. This will avoid the need for the type parameter. If we can eliminate all type params this way then we don't need the C#\VB specific types at all.

srivatsn replied:
`syntaxGenerator.GetDeclaration(node, DeclarationKind.ConversionOperator)` should avoid the need for the `TConversionOperatorSyntax`

srivatsn replied:
Looks like `DeclarationKind.Class`, `DeclarationKind.Method` and `DeclarationKind.Property` will remove the need for all type params and this class can be sealed and we can delete the C#\VB derived types.

srivatsn commented on src/Microsoft.ApiDesignGuidelines.Analyzers/Core/OperatorOverloadsHaveNamedAlternates.Fixer.cs:
```diff
@@ -27,10 +32,126 @@ public sealed override FixAllProvider GetFixAllProvider()
+            switch (diagnostic.Properties[OperatorOverloadsHaveNamedAlternatesAnalyzer.DiagnosticKindText])
+            {
+                case OperatorOverloadsHaveNamedAlternatesAnalyzer.AddAlternateText:
+                    AddFriendlyAlternate(context, root, node, semanticModel, diagnostic);
```
Registering the codefix inside this method after all computation is complete means that we'll do all that work just to see if we can produce a fix (which is a check that happens all the time during typing to see if a lightbulb needs to be shown). Instead, register a codefix at this location and have the codefix's callback do all the work only when it's invoked.

srivatsn commented on src/Microsoft.ApiDesignGuidelines.Analyzers/Core/OperatorOverloadsHaveNamedAlternates.Fixer.cs:
```diff
@@ -27,10 +32,126 @@ public sealed override FixAllProvider GetFixAllProvider()
+                    AddFriendlyAlternate(context, root, node, semanticModel, diagnostic);
+                    break;
+                case OperatorOverloadsHaveNamedAlternatesAnalyzer.FixVisibilityText:
+                    await FixFriendlyVisibility(context, node, semanticModel, diagnostic);
```
Same comment as above.

srivatsn commented on src/Microsoft.ApiDesignGuidelines.Analyzers/Core/OperatorOverloadsHaveNamedAlternates.Fixer.cs:
```diff
@@ -27,10 +32,126 @@ public sealed override FixAllProvider GetFixAllProvider()
+            var badVisibilitySymbol = semanticModel.GetDeclaredSymbol(badVisibilityNode, context.CancellationToken);
+            var symbolEditor = SymbolEditor.Create(context.Document);
+            var newSymbol = await symbolEditor.EditOneDeclarationAsync(badVisibilitySymbol,
+                (documentEditor, syntaxNode, ct) => Task.Run(() => documentEditor.SetAccessibility(badVisibilityNode, Accessibility.Public))).ConfigureAwait(false);
```
Minor: Do you need the Task.Run? I think there's a sync overload for `EditOneDeclarationAsync` as well.

srivatsn commented on src/Microsoft.ApiDesignGuidelines.Analyzers/Core/OperatorOverloadsHaveNamedAlternates.cs:
```diff
@@ -1,69 +1,262 @@
 
 namespace Microsoft.ApiDesignGuidelines.Analyzers
-{                   
+{
```
You can make this type sealed and delete the two derived type since they don't have any functionality. You'll need to add a `[DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic]` attribute to this type

srivatsn commented:
Thanks @brettfo  - LGTM modulo the comments.

brettfo commented:
Thanks for the pointers to `syntaxGenerator.GetDeclaration()`, that made the fixer a lot easier.

I've updated the PR with your feedback and removed the VB and C# specific code since it's all common now.

srivatsn commented on src/Microsoft.ApiDesignGuidelines.Analyzers/Core/OperatorOverloadsHaveNamedAlternates.Fixer.cs:
```diff
@@ -2,21 +2,24 @@
     /// CA2225: Operator overloads have named alternates
     /// </summary>
-    public abstract class OperatorOverloadsHaveNamedAlternatesFixer : CodeFixProvider
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic), Shared]
```
Wrong attribute - The fixer should have an ExportCodeFixProvider attribute

brettfo replied:
Thanks, copy/paste is hard :smile:.  I've updated the PR again.

srivatsn commented on src/Microsoft.ApiDesignGuidelines.Analyzers/Core/OperatorOverloadsHaveNamedAlternates.cs:
```diff
@@ -1,69 +1,264 @@
     /// CA2225: Operator overloads have named alternates
     /// </summary>
-    public abstract class OperatorOverloadsHaveNamedAlternatesAnalyzer : DiagnosticAnalyzer
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic), Shared]
```
DiagnosticAnalyzerAttribute is not a MEF attribute - so you don't need to add the Shared attribute here.

brettfo commented:
ping for other reviewers now that most of us are back from holidays

lgolding commented:
Hi @brettfo, I'll take a look now.

lgolding commented on src/Microsoft.ApiDesignGuidelines.Analyzers/Core/OperatorOverloadsHaveNamedAlternates.cs:
```diff
@@ -1,69 +1,263 @@
+            switch (operatorName)
+            {
+                case "op_True":
+                case "op_False":
```
`OpFalseText`, and introduce `OpTrueText`?

brettfo replied:
Done.

lgolding commented on src/Microsoft.ApiDesignGuidelines.Analyzers/Core/OperatorOverloadsHaveNamedAlternates.cs:
```diff
@@ -1,69 +1,263 @@
+            public ExpectedAlternateMethodGroup(string alternateMethod1, string alternateMethod2 = null)
+            {
+                AlternateMethod1 = alternateMethod1;
+                AlternateMethod2 = alternateMethod2;
```
Huh, learn something new every day. I didn't know this would work without the property having a `private set`.

brettfo replied:
Newly added in C# 6; read-only auto-properties are settable only from the constructor.

lgolding commented on src/Microsoft.ApiDesignGuidelines.Analyzers/Core/OperatorOverloadsHaveNamedAlternates.cs:
```diff
@@ -1,69 +1,263 @@
+            }
+        }
+
+        internal static ExpectedAlternateMethodGroup GetExpectedAlternateMethodGroup(string operatorName, ITypeSymbol returnType)
```
`private`?

lgolding commented on src/Microsoft.ApiDesignGuidelines.Analyzers/Core/OperatorOverloadsHaveNamedAlternates.cs:
```diff
@@ -1,69 +1,263 @@
+        private static void DoAnalysis(SymbolAnalysisContext symbolContext)
+        {
+            var methodSymbol = (IMethodSymbol)symbolContext.Symbol;
+            var location = methodSymbol.OriginalDefinition.Locations.First();
```
You don't need to calculate the location unless you're reporting certain of the diagnostics below. I'd wrap this in a helper method `LocationOfSymbol`, and then call `LocationOfSymbol(methodSymbol)` in each of the three places below where you report a diagnostic at `location`.

brettfo replied:
Done.

lgolding commented on src/Microsoft.ApiDesignGuidelines.Analyzers/Core/OperatorOverloadsHaveNamedAlternates.cs:
```diff
@@ -1,69 +1,263 @@
+            var methodSymbol = (IMethodSymbol)symbolContext.Symbol;
+            var location = methodSymbol.OriginalDefinition.Locations.First();
+            var typeSymbol = methodSymbol.ContainingSymbol as ITypeSymbol;
+            var operatorName = methodSymbol.Name;
```
I'd put this inside the `if` block, since you don't need it otherwise.

brettfo replied:
Done.

lgolding commented on src/Microsoft.ApiDesignGuidelines.Analyzers/Core/OperatorOverloadsHaveNamedAlternates.cs:
```diff
@@ -1,69 +1,263 @@
+                    }
+                    else if (!property.IsPublic())
+                    {
+                        symbolContext.ReportDiagnostic(CreateDiagnostic(VisibilityRule, property.OriginalDefinition.Locations.First(), FixVisibilityText, IsTrueText, operatorName));
```
If you had the helper method `LocationOfSymbol` I suggested above, you could using it here too, `LocationOfSymbol(property)`, and below, `LocationOfSymbol(notPublicMethod)`, which would DRY out the code a little.

brettfo replied:
Done.

lgolding commented on src/Microsoft.ApiDesignGuidelines.Analyzers/Core/OperatorOverloadsHaveNamedAlternates.cs:
```diff
@@ -1,69 +1,263 @@
+                        unmatchedMethods.Add(expectedGroup.AlternateMethod2);
+                    }
+
+                    foreach (var candidateMethod in typeSymbol.GetMembers().OfType<IMethodSymbol>())
```
You only need to do this once per enclosing type, not once per user-defined operator, but I supposed those are rare, so it's probably not a real perf issue.

brettfo replied:
I had your same concern, but I couldn't think of a nice way to cache the method symbols and I didn't want to go the other route of analyzing all types then looking for operator overloads, since the `if` block at the top of this method will filter out a lot of non-operator-overload symbols.

brettfo commented:
PR updated addressing @lgolding's comments.

lgolding commented on src/Microsoft.ApiDesignGuidelines.Analyzers/Core/OperatorOverloadsHaveNamedAlternates.cs:
```diff
@@ -1,69 +1,268 @@
+            }
+        }
+
+        private static Location GetSymbolLocation(ISymbol methodSymbol)
```
Nitpick: It's not always a method symbol, sometimes a property, so name s.b. just `symbol`.

brettfo replied:
Good catch, I'll update it.

lgolding commented on src/Microsoft.ApiDesignGuidelines.Analyzers/Core/OperatorOverloadsHaveNamedAlternates.Fixer.cs:
```diff
@@ -26,11 +28,123 @@ public sealed override FixAllProvider GetFixAllProvider()
+            switch (diagnostic.Properties[OperatorOverloadsHaveNamedAlternatesAnalyzer.DiagnosticKindText])
+            {
+                case OperatorOverloadsHaveNamedAlternatesAnalyzer.AddAlternateText:
+                    var methodDeclaration = generator.GetDeclaration(node, DeclarationKind.Operator) ?? generator.GetDeclaration(node, DeclarationKind.ConversionOperator);
```
`generator` might be null -- or at least, you used the null conditional operator `?.` above when you assigned it, so apparently you think it might be null. If so, we should do a null check before using it.

brettfo replied:
Done.

lgolding commented on src/Microsoft.ApiDesignGuidelines.Analyzers/Core/OperatorOverloadsHaveNamedAlternates.Fixer.cs:
```diff
@@ -26,11 +28,123 @@ public sealed override FixAllProvider GetFixAllProvider()
+            public IEnumerable<Tuple<string, ITypeSymbol>> Parameters { get; }
+            public bool IsStatic { get; }
+
+            public ExpectedMethodSignature(string name, ITypeSymbol returnType, IEnumerable<Tuple<string, ITypeSymbol>> parameters, bool isStatic = true)
```
It's not obvious to me that there is a reasonable default for `isStatic`. I'd remove the default and specify it explicitly everywhere.

brettfo replied:
Done.

lgolding commented on src/Microsoft.ApiDesignGuidelines.Analyzers/Core/OperatorOverloadsHaveNamedAlternates.Fixer.cs:
```diff
@@ -26,11 +28,123 @@ public sealed override FixAllProvider GetFixAllProvider()
+                    var operatorOverloadSymbol = (IMethodSymbol)semanticModel.GetDeclaredSymbol(methodDeclaration, context.CancellationToken);
+                    var typeSymbol = (ITypeSymbol)operatorOverloadSymbol.ContainingSymbol;
+                    var typeDeclarationSyntax = typeSymbol.DeclaringSyntaxReferences.First().GetSyntax(context.CancellationToken);
+                    var typeDeclaration = generator.GetDeclaration(typeDeclarationSyntax, DeclarationKind.Class);
```
It might also be a `DeclarationKind.Struct`.

lgolding replied:
In any case I'm confused by this whole sequence.You get the `operatorOverloadSymbol`, say, the symbol that declares the method `MyClass.operator +`. You get the containing type symbol (`MyClass`), then you get the syntax node `typeDeclarationSyntax` that declares the type (`class MyClass { ... }`). Then you call `SyntaxGenerator.GetDeclaration`, looking for either this node, or the nearest enclosing node, that declares a class. Why do you need this extra step? Why isn't `typeDeclarationSyntax` the node you want?

brettfo replied:
Re `struct`: the `GetDeclaration()` method also works for structs and I've added tests to verify this.

Re `typeDeclarationSyntax`: for C# these two nodes are the same, but for VB they're different and that's why this extra step is needed.  I'll add a comment explaining this here.

lgolding commented on src/Microsoft.ApiDesignGuidelines.Analyzers/Core/OperatorOverloadsHaveNamedAlternates.Fixer.cs:
```diff
@@ -26,11 +28,123 @@ public sealed override FixAllProvider GetFixAllProvider()
+
+                    SyntaxNode addedMember;
+                    var bodyStatements = ImmutableArray.Create(
+                        generator.ThrowStatement(generator.ObjectCreationExpression(semanticModel.Compilation.GetTypeByMetadataName("System.NotImplementedException"))));
```
Oh, this is cool, I haven't seen an example of this kind of code generation before.

heejaechang replied:
I think we have helper method for GetTypeByMetadataNamed(....)

lgolding commented on src/Microsoft.ApiDesignGuidelines.Analyzers/Core/OperatorOverloadsHaveNamedAlternates.Fixer.cs:
```diff
@@ -26,11 +28,123 @@ public sealed override FixAllProvider GetFixAllProvider()
+                case "op_UnaryNegation":
+                case "op_UnaryPlus":
+                    // e.g., public MyClass Decrement(MyClass item)
+                    return new ExpectedMethodSignature(expectedName, returnType, ImmutableArray.Create(Tuple.Create("item", containingType)));
```
`isStatic` is `true` by default. Is the comment wrong? That is, should the example be `public static MyClass Decrement(MyClass item)`?

brettfo replied:
Yes, the comment is wrong; good catch, I'll update it.  And I agree with your `isStatic` comment elsewhere and I've fixed that, too.

lgolding commented on src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/OperatorOverloadsHaveNamedAlternatesTests.Fixer.cs:
```diff
@@ -10,24 +9,402 @@ namespace Microsoft.ApiDesignGuidelines.Analyzers.UnitTests
+
+    public static C Add(C left, C right)
+    {
+        throw new System.NotImplementedException();
```
This is way beyond the MVP for this rule, but... Why not return the result of calling the operator?

```
public static C Add(C left, C right)
{
    return left + right;
}
```

brettfo replied:
I'd considered this, but ultimately decided against it because not all operators have such obvious defaults.  E.g., if this fixer generates a `ToInt32()` method, how does that `int` get generated?  Because of this I decided to stick with the known pattern of throwing.

brettfo commented:
Updated with latest changes.

lgolding commented:
:+1:

heejaechang commented on src/Microsoft.ApiDesignGuidelines.Analyzers/Core/OperatorOverloadsHaveNamedAlternates.Fixer.cs:
```diff
@@ -2,21 +2,23 @@
 using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
+using Analyzer.Utilities;
```
sort?

brettfo replied:
I re-sorted with `System` namespaces first.

heejaechang commented on src/Microsoft.ApiDesignGuidelines.Analyzers/Core/OperatorOverloadsHaveNamedAlternates.Fixer.cs:
```diff
@@ -26,11 +28,130 @@ public sealed override FixAllProvider GetFixAllProvider()
-            throw new NotImplementedException();
+        public sealed override Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            return Task.Run(() => context.RegisterCodeFix(new MyCodeAction(ct => Fix(context, ct)), context.Diagnostics.First()));
```
I think either you want to do

context.ReigsterCodeFix ..
return Task.FromResult<object>(null);

or actually do some checking in this method.

heejaechang commented on src/Microsoft.ApiDesignGuidelines.Analyzers/Core/OperatorOverloadsHaveNamedAlternates.Fixer.cs:
```diff
@@ -2,21 +2,23 @@
     /// CA2225: Operator overloads have named alternates
     /// </summary>
-    public abstract class OperatorOverloadsHaveNamedAlternatesFixer : CodeFixProvider
+    [ExportCodeFixProvider(LanguageNames.CSharp, LanguageNames.VisualBasic), Shared]
```
do we actually need this MEF export? I though we do attribute based search for fixer as well.

heejaechang commented on src/Microsoft.ApiDesignGuidelines.Analyzers/Core/OperatorOverloadsHaveNamedAlternates.Fixer.cs:
```diff
@@ -26,11 +28,130 @@ public sealed override FixAllProvider GetFixAllProvider()
+
+        private static async Task<Document> Fix(CodeFixContext context, CancellationToken cancellationToken)
+        {
+            var semanticModel = await context.Document.GetSemanticModelAsync(context.CancellationToken).ConfigureAwait(false);
```
so this kind of stuff is what you can do in the RegisterCodeFixAsync so that we dont show lightbulb but when try to execute it, it gets removed from the fixer list.

brettfo replied:
Good call.  I'll move this and the `SyntaxGenerator` stuff to `RegisterCodeFixesAsync()` and only call `context.RegisterCodeFix` if those values are non-null.

heejaechang commented on src/Microsoft.ApiDesignGuidelines.Analyzers/Core/OperatorOverloadsHaveNamedAlternates.Fixer.cs:
```diff
@@ -26,11 +28,130 @@ public sealed override FixAllProvider GetFixAllProvider()
+                case OperatorOverloadsHaveNamedAlternatesAnalyzer.AddAlternateText:
+                    var methodDeclaration = generator.GetDeclaration(node, DeclarationKind.Operator) ?? generator.GetDeclaration(node, DeclarationKind.ConversionOperator);
+                    var operatorOverloadSymbol = (IMethodSymbol)semanticModel.GetDeclaredSymbol(methodDeclaration, context.CancellationToken);
+                    var typeSymbol = (ITypeSymbol)operatorOverloadSymbol.ContainingSymbol;
```
I think there is containingType method

brettfo replied:
Good catch, thanks.

heejaechang commented on src/Microsoft.ApiDesignGuidelines.Analyzers/Core/OperatorOverloadsHaveNamedAlternates.Fixer.cs:
```diff
@@ -26,11 +28,130 @@ public sealed override FixAllProvider GetFixAllProvider()
+            {
+                case OperatorOverloadsHaveNamedAlternatesAnalyzer.AddAlternateText:
+                    var methodDeclaration = generator.GetDeclaration(node, DeclarationKind.Operator) ?? generator.GetDeclaration(node, DeclarationKind.ConversionOperator);
+                    var operatorOverloadSymbol = (IMethodSymbol)semanticModel.GetDeclaredSymbol(methodDeclaration, context.CancellationToken);
```
does this work? the method declaration is what you just created, and symantic model is from original document. semantic model will not have the method decl node you just created.

heejaechang commented on src/Microsoft.ApiDesignGuidelines.Analyzers/Core/OperatorOverloadsHaveNamedAlternates.Fixer.cs:
```diff
@@ -26,11 +28,130 @@ public sealed override FixAllProvider GetFixAllProvider()
+                    var badVisibilitySymbol = semanticModel.GetDeclaredSymbol(badVisibilityNode, context.CancellationToken);
+                    var symbolEditor = SymbolEditor.Create(context.Document);
+                    var newSymbol = await symbolEditor.EditOneDeclarationAsync(badVisibilitySymbol,
+                        (documentEditor, syntaxNode, ct) => Task.Run(() => documentEditor.SetAccessibility(badVisibilityNode, Accessibility.Public))).ConfigureAwait(false);
```
same as aboive. I think you should do return Task.FromResult... rather than asking it to actually run task.

brettfo replied:
Done.

brettfo commented:
I've opened #784 to address @heejaechang's feedback.
